docs/advanced/UsageWithTypescript.md

```diff
@@ -0,0 +1,219 @@
+
+export interface ISystemState {
+  loggedIn: boolean
+  session: string
+  userName: string
+}
+```
+
+Note that we are exporting these interfaces to reuse them later in application state and action creators.
+
+We can now create an interface which describes the global store:
+
+```ts
+// src/store/index.ts
+
+export interface IAppState {
+  system: ISystemState
+  chat: IChatState
+}
+```
+
+## Type Checking Actions & Action Creators
+
+We will be using TypeScript's enums to declare our action constants. [Enums](https://www.typescriptlang.org/docs/handbook/enums.html) allow us to define a set of named constants.
+
+Chat Action Constants and Actions:
+
+```ts
+// src/store/chat/types.ts
+
+export enum ChatActions {
+  SendMessage = '@CHAT:SEND/MESSAGE'
+}
+
+export interface ISendMessageAction {
+  type: ChatActions.SendMessage
+  payload: IMessage
+}
+```
+
+With these types we can now also type check chat's action creators:
+
+```ts
+// src/store/chat/actions.ts
+
+import { IMessage, ISendMessageAction, ChatActions } from './types'
+
+export type sendMessageType = (newMessage: IMessage) => ISendMessageAction
+
+export const sendMessage: sendMessageType = newMessage => ({
+  type: ChatActions.SendMessage,
+  payload: newMessage
+})
+```
+
+System Action Constants and Actions:
+
+```ts
+// src/store/system/types.ts
+
+export enum SystemActions {
+  UpdateSession = '@SYSTEM:UPDATE/SESSION'
+}
+
+export interface IUpdateSessionAction {
+  type: SystemActions.UpdateSession
+  payload: ISystemState
+}
+```
+
+With these types we can now also type check system's action creators:
+
+```ts
+// src/store/system/actions.ts
+
+import { SystemActions, IUpdateSessionAction, ISystemState } from './types'
+
+export type updateSessionType = (
+  newSession: ISystemState
+) => IUpdateSessionAction
+
+export const updateSession: updateSessionType = newSession => ({
+  type: SystemActions.UpdateSession,
+  payload: newSession
+})
+```
+
+## Type Checking Reducers
+
+Reducers are just pure functions that take the previous state, an action and then return the next state. Redux provides a `Reducer` type which can be imported to assist with type checking.
+
+Type checked chat reducer:
+
+```ts
+// src/store/chat/reducers.ts
+
+import { Reducer } from 'redux'
+import { IChatState, ChatActions } from './types'
+
+const initialState: IChatState = {
+  messages: []
+}
+
+export const chatReducer: Reducer<IChatState> = (
+  state = initialState,
+  action
+) => {
+  switch (action.type) {
+    case ChatActions.SendMessage: {
+      return {
+        messages: [...state.messages, action.payload]
+      }
+    }
+    default:
+      return state
+  }
+}
+```
+
+Type checked system reducer:
+
+```ts
+// src/store/system/reducers.ts
+
+import { Reducer } from 'redux'
+import { SystemActions, ISystemState } from './types'
+
+const initialState: ISystemState = {
+  loggedIn: false,
+  session: '',
+  userName: ''
+}
+
+export const systemReducer: Reducer<ISystemState> = (
```

Comment:
This is more of a criticism of the code than the TypeScript: IMO it's best practice to have one reducer manage *one thing*. Instead of having a state object with multiple values, you should have each reducer manage one part of each state unless that state is instrinically tied together (in most cases, state that is that tightly coupled can be derived state).

For example, `loggedIn: false` could be determined by whether or not `session` is equal to `null`. `session` should be `null` as should `userName` with absent values.

Just my 2c - this could be changed to be two reducers, which would help keep the state flat - which is important in Redux!

```ts
export function sessionReducer(opaqueSessionToken: string | null, action: Action) {
  ...
}

export function usernameReducer(username: string | null, action: Action) {
  ...
}


// src/store/index.ts
const rootReducer = combineReducers({
  user: userReducer,
  session: sessionReducer,
  chat: chatReducer
})
```